src/relay/qnn/pass/qnn_lower.cc

```diff
@@ -0,0 +1,271 @@
+ *
+ *       Credit to TFLite reference implementation.
+ */
+std::pair<int, int> GetFixedPointMultiplierShift(double double_multiplier,
+    const DataType& idtype) {
+  int significand, exponent;
+  if (double_multiplier == 0.) {
+    significand = 0;
+    exponent = 0;
+    return std::pair<int, int>(significand, exponent);
+  }
+  int idtype_bits = idtype.bits();
+
+  // Get the significand (significand) and exponent (exponent)
+  double significand_d = std::frexp(double_multiplier, &exponent);
+
+  // Convert the double significand to int significand.
+  significand_d = std::round(significand_d * (1ll << (idtype_bits - 1)));
+  auto significand_int64 = static_cast<int64_t>(significand_d);
+  CHECK_LE(significand_int64, (1ll << (idtype_bits - 1)));
+  if (significand_int64 == (1ll << (idtype_bits - 1))) {
+    significand_int64 /= 2;
+    ++exponent;
+  }
+  CHECK_LE(significand_int64, std::numeric_limits<int>::max());
+  significand = static_cast<int>(significand_int64);
+  return std::pair<int, int>(significand, exponent);
+}
+
+/*
+ * \brief Lower requantize to a sequence of ops.
+ * \param input_tensor The input tensor to requantize op.
+ * \param param The requantize op attrs.
+ * \param idtype The dtype of the input tensor.
+ * \param out_shape The output shape of the requantize op.
+ * \return The sequence of existing Relay ops.
+ * \note Requantization using only integer computation. Here, the computation is
+ *       converted to a fixed point computation by computing output multiplier
+ *       and shift. This is useful, if the target device does not support/have
+ *       very expensive floating point computations.
+ *
+ *       Original compuation is scale_fp32 * quantized_tensor.  To convert into
+ *       integer computation, the multiplication with fp32 scalar can be
+ *       replaced by multiplication with an int value and then right shifting
+ *       the result. This approximates the floating point computation with a
+ *       fixed point computation.
+ *
+ *       The whole computation this can be broken down into following steps
+ *       1) Calculate the integer multiplier and integer shift.
+ *       2) Subtract the input integer point.
+ *       3) Multiply the integer fixed point multiplier with quantized tensor.
+ *       4) Round the result.
+ *       5) Right shift the result.
+ *       6) Add the output_zero_point.
+ *       7) Cast to the out_dtype.
+ */
+Expr RequantizeLower(const Expr& input_tensor,
+    const RequantizeAttrs* param, const DataType& idtype,
+    const Array<IndexExpr>& out_shape) {
+
+  double double_multiplier = param->input_scale/param->output_scale;
+
+  // The multiplication will be performed in higher precision. Find the dtype.
+  int idtype_bits = idtype.bits();
+  DataType up_idtype = Int(2 * idtype_bits);
+
+  // 1) Calculating the integer multiplier and integer shift
+  std::pair<int, int> fixed_point_params =
+      GetFixedPointMultiplierShift(double_multiplier, idtype);
```

Comment:
I see this `idtype` could be of signed/unsigned integer of 8/16/32 bits. They could be safely handled by an `int32_t` representation, so are don't need the `idtype` parameter for `GetFixedPointMultiplierShift()`?

Reply:
Thanks @jackwish for raising this point. Earlier I was trying to do minimal up_idtype. For example, `int8` input can just go to `int16`. But, your point that we can handle everything in `int32_t` is stronger.

I am changing the code to remove the dependence on in_dtype. We will always go to int32_t for any in_dtype. This will make code shorter and thus less prone to bugs. I already have tests with `int8`, so I can check the functionality as well.

Thanks for looking into such a detail :)